ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -847,21 +833,32 @@ let start () =
   in
   let configuration = Configuration.default in
   let detached = Fiber.Pool.create () in
-  let dune = Dune.create scheduler in
+  let server = Fdecl.create Dyn.Encoder.opaque in
+  let diagnostics =
+    Diagnostics.create (fun diagnostics ->
+        let server = Fdecl.get server in
+        let state = Server.state server in
+        task_if_running state ~f:(fun () ->
+            Server.notification server (PublishDiagnostics diagnostics)))
+  in
+  let dune = Dune.create scheduler diagnostics in
   let server =
     let merlin = Scheduler.create_thread scheduler in
     let ocamlformat = Fmt.create scheduler in
-    Server.make handler stream
-      { store
-      ; init = Uninitialized
-      ; merlin
-      ; ocamlformat
-      ; scheduler
-      ; configuration
-      ; detached
-      ; dune
-      ; trace = `Off
-      }
+    Fdecl.set server
+      (Server.make handler stream
```

Comment:
Why use a forward declaration when you're setting the value in the same function? `server` could also be passed to `Diagnostics.create` to avoid having to define a `send` outside of `Diagnostics` as commented above.

Reply:
It's done to break the following cycle:

* `dune` depends on `diagonstics`
* `server` depends on `dune`
* `diagnostics` depends on `server`

Perhaps it's possible to avoid this cycle with `lazy` and `let .. and`, but it's more work because some of the bindings require fibers.

Reply:
Understand, thanks!